### PR TITLE
fix: reduce memory usage in ingester with less buffering prior to query engine

### PR DESCRIPTION
**Pull request by alamb**
Here are the pieces related to this PR:
- [x] https://github.com/influxdata/influxdb_iox/pull/4823
- [x] https://github.com/influxdata/influxdb_iox/pull/4822
- [x] https://github.com/influxdata/influxdb_iox/pull/4832
- [x] https://github.com/influxdata/influxdb_iox/pull/4837
- [x] https://github.com/influxdata/influxdb_iox/pull/4855

# Rationale
Related to https://github.com/influxdata/influxdb_iox/issues/4567

The ingester buffers data twice prior to sending it to the ingester as illustrated in the following diagram.

```text
                  Buffering location             Buffering location #2:
                  #1: turned into a               turned into a single
┌─────────────┐   single RecordBatch              RecordBatch prior to
│┌───────────┐│   to feed into query               sending to querier
││RecordBatch││         engine
│└───────────┘├─────┐
│             │     │
│SnapshotBatch│     │      ┌────────────────────────┐       ┌ ─ ─ ─ ─ ─ ┐        ┌────────────────────────┐
└─────────────┘     │      │                        │                            │                        │
                    ├─────▶│      Query Engine      │──────▶│    ...    │────┬──▶│ IngesterQueryResponse  │
      ...           │      │                        │                        │   │                        │
                    │      └────────────────────────┘       └ ─ ─ ─ ─ ─ ┘    │   └────────────────────────┘
┌─────────────┐     │                                                        │
│┌───────────┐│     │     Runs query engine, which                           │
││RecordBatch││     │     deduplicates and applies                           │
│└───────────┘│─────┘    filters, produces a Stream                          │
│             │                                                            ──┤
│SnapshotBatch│                                                              │
└─────────────┘                                                              │
                                                                     ...     │
                                                                             │
                                                                             │
                                                            data from other  │
                                                               partitions    │
                                                                          ───┘

```

# Changes

Removes a redundant filtering and combination, as illustrated below

```
               ▲                                                 
               │                                                 
               │                                                 
               │                                                 
               │                                                 
               │                                                 
               │                                                 
┌────────────────────────────┐                                   
│           Filter           │                                   
│   (evaluates prediates)    │                                   
└────────────────────────────┘                                   
               ▲                                                 
               │                                                 
               │                                                 
               │                                                 
 ┌───────────────────────────┐                                   
 │           Scan            │                                   
 │    (calls read_filter)    │◀─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┐           
 └───────────────────────────┘                                   
                                                     │           
              Plan                         Applying predicates   
                                          during read_filter is  
                                          unecessary as they are 
                                           applied immediately   
                                           again in the Filter   
```

### CODE REVIEW

File: ingester/src/query.rs [hunk 1]
```diff
@@ -204,76 +197,45 @@ impl QueryChunk for QueryableBatch {
...
-                    ?num_rows_after,
-                    "predicate pushdown for QueryableBatch"
-                );
-            }
-
-            if batch.num_rows() > 0 {
-                stream_batches.push(Arc::new(batch));
-            }
-        }
+        // Get all record batches from their snapshots
```
Comment (alamb) [inline ingester/src/query.rs:211]: This change stops applying an extra level of filtering / squashing in the `Scan` -- the `Filter` node in the actual query plan does the filtering again so applying it in the scan is unnecessary and no faster than what the Filter will do.

```
               ▲                                                 
               │                                                 
               │                                                 
               │                                                 
               │                                                 
               │                                                 
               │                                                 
┌────────────────────────────┐                                   
│           Filter           │                                   
│   (evaluates prediates)    │                                   
└────────────────────────────┘                                   
               ▲                                                 
               │                                                 
               │                                                 
               │                                                 
 ┌───────────────────────────┐                                   
 │           Scan            │                                   
 │    (calls read_filter)    │◀─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┐           
 └───────────────────────────┘                                   
                                                     │           
              Plan                         Applying predicates   
                                          during read_filter is  
                                          unecessary as they are 
                                           applied immediately   
                                           again in the Filter   
```

File: ingester/src/query.rs [hunk 2]
```diff
@@ -204,76 +197,45 @@ impl QueryChunk for QueryableBatch {
...
+        let schema = self
+            .schema()
+            .select(selection)
+            .context(SchemaSnafu)?;
 
-        // Combine record batches into one batch and padding null values as needed
-        // Schema of all record batches after mergeing
-        let schema = merge_record_batch_schemas(&batches);
-        let batch =
-            merge_record_batches(schema.as_arrow(), batches).context(ConcatBatchesSnafu {})?;
```
Comment (alamb) [inline ingester/src/query.rs:249]: this `merge_record_batches` is a copy and is removed as part of this PR

File: ingester/src/query.rs [hunk 3]
```diff
@@ -384,241 +343,6 @@ mod tests {
         assert_eq!(expected, predicates);
     }
 
-    #[tokio::test]
-    async fn test_read_filter() {
```
Comment (alamb) [inline ingester/src/query.rs:394]: Correct answers are already verified in:
1. `query_tests` (❤️  to @crepererum  for wiring that up to use the actual `IngesterResponse`)
2. the `querier_handler` tests, such as https://github.com/influxdata/influxdb_iox/blob/main/ingester/src/querier_handler.rs#L391-L393

File: arrow_util/src/util.rs [hunk 4]
```diff
@@ -23,50 +23,55 @@ where
     RecordBatch::try_from_iter(vec![(field_name, Arc::new(array) as ArrayRef)])
 }
 
+/// Ensures the record batch has the specified
```
Comment (alamb) [inline arrow_util/src/util.rs:26]:
```suggestion
/// Ensures the record batch has the specified schema
```

File: arrow_util/src/util.rs [hunk 5]
```diff
@@ -23,50 +23,55 @@ where
...
+pub fn ensure_schema(
+    output_schema: &SchemaRef,
+    batch: &RecordBatch,
+) -> Result<RecordBatch, ArrowError> {
+    let batch_schema = batch.schema();
+
+    // Go over all columns of output_schema
```
Comment (alamb) [inline arrow_util/src/util.rs:33]: this simply refactors the logic to extend the schema for a batch to its own function

File: arrow_util/src/util.rs [hunk 6]
```diff
@@ -23,50 +23,55 @@ where
...
-                        new_null_array(output_field.data_type(), batch.num_rows())
-                    }
-                })
-                .collect::<Vec<_>>();
-
-            RecordBatch::try_new(Arc::clone(&output_schema), batch_output_columns)
-                .expect("A record batch should have been created")
-        })
-        .collect::<Vec<_>>();
+        .map(|batch| ensure_schema(output_schema, batch.as_ref()))
```
Comment (NGA-TRAN) [inline arrow_util/src/util.rs:68]: 👍

File: ingester/src/query.rs (hunk 2 above)
Comment (NGA-TRAN) [reply ingester/src/query.rs:249]: Agree

File: ingester/src/query.rs (hunk 1 above)
Comment (NGA-TRAN) [reply ingester/src/query.rs:211]: Nice

File: ingester/src/query.rs (hunk 3 above)
Comment (NGA-TRAN) [reply ingester/src/query.rs:394]: Hope these are all covered there 😄

**Review by NGA-TRAN — APPROVED** [review]

**Pull request merged**